packages/expo-font/android/src/main/java/expo/modules/font/FontLoaderModule.java

```diff
@@ -6,71 +6,118 @@
-        return;
+      Typeface typeface = getTypeface(localUri);
+      if (typeface == null) {
+        throw new FontFileInvalidException(localUri);
       }
-      fontManager.setTypeface(prefix + fontFamilyName, Typeface.NORMAL, typeface);
+
+      ReactFontManager.getInstance().setTypeface(fontFamilyName, Typeface.NORMAL, typeface);
+      mLoadedFontNames.add(fontFamilyName);
       promise.resolve(null);
-    } catch (Exception e) {
-      promise.reject("E_UNEXPECTED", "Font.loadAsync unexpected exception: " + e.getMessage(), e);
+    } catch (CodedRuntimeException e) {
+      // Most probably an InvalidArgumentException. Already coded!
+      promise.reject(e);
+    } catch (RuntimeException e) {
+      // Runtime exception is thrown if and only if there's no font file
+      promise.reject(new FontFileNotFoundException(fontFamilyName, localUri));
     }
   }
 
-  @Override
-  public void onCreate(ModuleRegistry moduleRegistry) {
-    mModuleRegistry = moduleRegistry;
+  protected Typeface getTypeface(String localUri) throws InvalidArgumentException {
+    if (localUri.startsWith(ASSET_SCHEME)) {
+      return Typeface.createFromAsset(
+        getContext().getAssets(),
+        // Also remove the leading slash.
+        localUri.substring(ASSET_SCHEME.length() + 1));
+    }
+
+    String localFontPath = Uri.parse(localUri).getPath();
+    if (localFontPath == null) {
+      throw new InvalidArgumentException("Could not parse provided local font URI as a URI with a path component.");
+    }
+    return Typeface.createFromFile(new File(localFontPath));
+  }
+
+  protected static class FontFileNotFoundException extends CodedRuntimeException {
+    public FontFileNotFoundException(String fontFamilyName, String path) {
+      super(String.format("File '%s' for font '%s' doesn't exist.", path, fontFamilyName));
+    }
+
+    @Override
+    public String getCode() {
+      return "ERR_FONT_FILE_NOT_FOUND";
+    }
   }
 
-  private boolean isScoped() {
-    ConstantsInterface constantsModule = mModuleRegistry.getModule(ConstantsInterface.class);
-    // If there's no constants module, or app ownership isn't "expo", we're not in Expo Client.
-    return constantsModule != null && "expo".equals(constantsModule.getAppOwnership());
+  protected static class FontFileInvalidException extends CodedRuntimeException {
+    public FontFileInvalidException(String path) {
+      super(String.format("File '%s' isn't a valid font file.", path));
+    }
+
+    @Override
+    public String getCode() {
+      return "ERR_FONT_FILE_INVALID";
+    }
+  }
+
+  protected static class FontAlreadyLoadedException extends CodedRuntimeException {
+    public FontAlreadyLoadedException(String fontFamilyName) {
+      super(String.format("Font with family name '%s' has already been loaded.", fontFamilyName));
+    }
+
+    @Override
+    public String getCode() {
+      return "ERR_FONT_ALREADY_LOADED";
+    }
```

Comment:
Extract to separate file 🙏 
Moreover this boilerplace is killing me 😞
I'd like to gradually move to kotlin whenever possible (every time a new chunk of code is created I wish it is kotlin and not java).